EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFUnpackerOutput.cc

```diff
@@ -6,72 +6,100 @@
+	return true;
+      }
+
+      LogDebug("L1T") << "nBX = " << nBX << " firstBX = " << firstBX << " lastBX = " << lastBX;
 			
-			int processor = block.amc().getBoardID() - 1;
-			if ( processor < 0 || processor > 11 )
-			{
-				edm::LogInfo ("l1t:stage2::BMTFUnpackerOutput::unpack") << "Processor found out of range so it will be calculated by the old way";
-				if ( block.amc().getAMCNumber()%2 != 0 )
-			                processor = block.amc().getAMCNumber()/2 ;
-			        else
-			                processor = 6 + (block.amc().getAMCNumber()/2 -1);
-			}
+      int processor = block.amc().getBoardID() - 1;
+      if ( processor < 0 || processor > 11 )
+	{
+	  edm::LogInfo ("l1t:stage2::BMTFUnpackerOutput::unpack") << "Processor found out of range so it will be calculated by the old way";
+	  if ( block.amc().getAMCNumber()%2 != 0 )
+	    processor = block.amc().getAMCNumber()/2 ;
+	  else
+	    processor = 6 + (block.amc().getAMCNumber()/2 -1);
+	}
 
-			for(int ibx = firstBX; ibx <= lastBX; ibx++)
-			{
-				int ip(0);
-				for(unsigned int iw = 0; iw < block.header().getSize()/nBX; iw += 2)
-				{
-					uint32_t raw_first = payload[ip+(ibx+lastBX)*6];
-					ip++;
-					uint32_t raw_secnd = payload[ip+(ibx+lastBX)*6];
-					ip++;
-					if ( raw_first == 0 )
-					{
-						LogDebug("L1T") << "Raw data is zero";
-						continue;
-					}
-					
-					RegionalMuonCand muCand = RegionalMuonCand();
-					RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf);
-					muCand.setLink(48 + processor);	//the link corresponds to the uGMT input
+      for (auto bxBlock : bxBlocks) {
+	int ibx = bxBlock.header().getBx();
 
-					LogDebug("L1T") << "Pt = " << muCand.hwPt() << " eta: " << muCand.hwEta() << " phi: " << muCand.hwPhi();
-					if ( muCand.hwQual() != 0 )
-					{
-						res->push_back(ibx, muCand);
-					}
+	for (auto iw = 0; iw < 6; iw+=2) {
+	  uint32_t raw_first = bxBlock.payload()[iw];//payload[ip+(ibx+lastBX)*6];
+	  uint32_t raw_secnd = bxBlock.payload()[iw+1];//payload[ip+(ibx+lastBX)*6];
+	  if ( raw_first == 0 )
+	    {
+	      LogDebug("L1T") << "Raw data is zero";
+	      continue;
+	    }
 					
-				}//for iw
-			}//for ibx
+	  RegionalMuonCand muCand;
+	  RegionalMuonRawDigiTranslator::fillRegionalMuonCand(muCand, raw_first, raw_secnd, processor, tftype::bmtf);
+
+	  if (isKalman) {
+	    //muCand.setLink(????????); to be filled later
+	    muCand.setHwPt2((raw_secnd >> 23) & 0xFF);
+	    muCand.setHwDXY((raw_secnd >> 18) & 0x3);
+	    LogDebug("L1T") << "Pt = " << muCand.hwPt() << " eta: " << muCand.hwEta() << " phi: " << muCand.hwPhi() << " diplacedPt = " << muCand.hwPt2();
+	  }
+	  else {
+	    muCand.setLink(48 + processor);	//the link corresponds to the uGMT input
+	    LogDebug("L1T") << "Pt = " << muCand.hwPt() << " eta: " << muCand.hwEta() << " phi: " << muCand.hwPhi();
+	  }
+
+	  if ( muCand.hwQual() != 0 ) {
```

Comment:
If you check this right after making the RegionalMuonCand you can save some time when hwQual == 0.